ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/sink/config/SinkConfig.java

```diff
@@ -240,15 +239,18 @@ Output getOutput(Env env) {
             env.getDuration(STREAMING_BATCH_MAX_DELAY, DEFAULT_STREAMING_BATCH_MAX_DELAY),
             PubsubMessageToTemplatedString.forBigQuery(env.getString(OUTPUT_TABLE)),
             getFormat(env));
-        long maxStreamingSize = env.getLong(STREAMING_MESSAGE_MAX_BYTES,
-            DEFAULT_STREAMING_MESSAGE_MAX_BYTES);
-        return new Output(env, this, message -> {
-          if (message.getSerializedSize() > maxStreamingSize) {
-            return fileOutput.apply(message);
-          } else {
-            return streamingOutput.apply(message);
-          }
-        });
+        Function<PubsubMessage, CompletableFuture<Void>> mixedOutput = message -> streamingOutput
+            .apply(message).thenApply(CompletableFuture::completedFuture).exceptionally(t -> {
+              if (t.getCause() instanceof BigQuery.WriteErrors) {
+                BigQuery.WriteErrors cause = (BigQuery.WriteErrors) t.getCause();
+                if (cause.errors.size() == 1 && cause.errors.get(0).getMessage()
+                    .startsWith("Maximum allowed row size exceeded")) {
+                  return fileOutput.apply(message);
+                }
+              }
+              throw (RuntimeException) t;
+            }).thenCompose(v -> v);
+        return new Output(env, this, mixedOutput);
```

Comment:
> Row size is calculated for the encoded row, which is within a few bytes of protobuf serialized size, so the default is 1MB to leave room for differences.

@jklukas I rewrote this to use the file output only if BigQuery actually rejects the message, because:

- the encoded row for `payload` format is usually much smaller than the JSON
- the live sink is the only one currently using `mixed` mode
- we are limited 1,000 or fewer messages per table via `fileOutput` in `mixed` mode, because every worker submits their own load jobs

downsides of this change include:
- there is a reference to `message` kept until it is successfully delivered to `streamingOutput`, so it can't be garbage collected
  - In practice this doesn't seem to have any impact on memory consumption, so I suspect it's not the only lingering reference.
- messages sent to `fileOutput` go through `PubsubMessageToObjectNode` twice
  - we could apply that as a transform, but it's a non-trivial change that I don't want to make that change as part of this PR

Reply:
Glancing through the BigQuery code, it looks like we're able to reasonable efficiently map individual messages to their associated errors. This looks reasonable. It would be good to send a 2 MB message and test that this all behaves as we expect in practice.

Reply:
> It would be good to send a 2 MB message and test that this all behaves as we expect in practice.

already did ^_^ https://github.com/mozilla/gcp-ingestion/pull/1124/files#diff-a8c5a168b5959568d77819c522c24427R14-R34